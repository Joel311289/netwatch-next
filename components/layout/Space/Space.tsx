import React from 'react';
import cx from 'classnames';
import { ElementChildren, ElementHTML } from '@/types';
import styles from '@/components/layout/Space/Space.module.scss';

export const positions = {
  center: 'center',
  start: 'flex-start',
  end: 'flex-end',
  between: 'space-between',
  baseline: 'baseline'
};

export type DirectionSpace = 'row' | 'column';

export type PositionSpace = keyof typeof positions;

export type SpaceProps = typeof defaultProps &
  ElementHTML &
  ElementChildren & {
    direction?: DirectionSpace;
    reverse?: boolean;
    wrap?: boolean;
    gap?: number | Array<number>;
    justify?: PositionSpace;
    align?: PositionSpace;
  };

const defaultProps = {
  direction: 'row'
};

const Space = ({
  className,
  children,
  style,
  direction,
  reverse,
  wrap,
  gap,
  justify,
  align
}: SpaceProps) => {
  const abc = 123;

  const spacing = React.useMemo(
    () =>
      gap
        ? [gap]
            .flat()
            .map((s: number) => `${s}px`)
            .join(' ')
        : '',
    [gap]
  );

  return (
    <div className={cx(styles.wrapper, className)} style={style}>
      {children}

      <style jsx>{`
        .${styles.wrapper} {
          display: flex;
          flex-direction: ${reverse ? `${direction}-reverse` : direction};
          flex-wrap: ${wrap ? 'wrap' : 'nowrap'};
          ${gap ? `gap: ${spacing}` : ''}
          ${justify ? `justify-content: ${positions[justify]}` : ''}
          ${align ? `align-items: ${positions[align]}` : ''}
        }
      `}</style>
    </div>
  );
};

Space.defaultProps = defaultProps;

export default Space;
