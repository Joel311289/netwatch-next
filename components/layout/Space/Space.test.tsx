import { render, screen } from '@testing-library/react';
import Space, { positions, DirectionSpace, PositionSpace } from '@/components/layout/Space/Space';

describe('Tests Space component', () => {
  test('renders component correctly', () => {
    const { container } = render(<Space>Test</Space>);

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    render(<Space>Test</Space>);

    expect(screen.getByText('Test')).toBeInTheDocument();
  });

  (['row', 'column'] as Array<DirectionSpace>).forEach((direction) => {
    test(`renders with style "flex-direction: ${direction}" with direction prop to ${direction}`, () => {
      render(<Space direction={direction}>Test</Space>);

      expect(screen.getByText('Test')).toHaveStyle(`flex-direction: ${direction}`);
    });
  });

  (['row', 'column'] as Array<DirectionSpace>).forEach((direction) => {
    test(`renders with style "flex-direction: ${direction}-reverse" with direction prop to ${direction} and reverse prop to true`, () => {
      render(
        <Space direction={direction} reverse>
          Test
        </Space>
      );

      expect(screen.getByText('Test')).toHaveStyle(`flex-direction: ${direction}-reverse`);
    });
  });

  (Object.keys(positions) as Array<PositionSpace>).forEach((justify) => {
    test(`renders with style "justify-content: ${positions[justify]}" with justify prop to ${justify}`, () => {
      render(<Space justify={justify}>Test</Space>);

      expect(screen.getByText('Test')).toHaveStyle(`justify-content: ${positions[justify]}`);
    });
  });

  (Object.keys(positions) as Array<PositionSpace>).forEach((align) => {
    test(`renders with style "align-items: ${positions[align]}" with align prop to ${align}`, () => {
      render(<Space align={align}>Test</Space>);

      expect(screen.getByText('Test')).toHaveStyle(`align-items: ${positions[align]}`);
    });
  });

  ['nowrap', 'wrap'].forEach((wrap, index) => {
    test(`renders with style "flex-wrap: ${wrap}" with wrap prop to ${Boolean(index)}`, () => {
      render(<Space wrap={Boolean(index)}>Test</Space>);

      expect(screen.getByText('Test')).toHaveStyle(`flex-wrap: ${wrap}`);
    });
  });

  test('renders with style "gap" with gap prop', () => {
    const gap = 5;
    render(<Space gap={gap}>Test</Space>);

    expect(screen.getByText('Test')).toHaveStyle(`gap: ${gap}px`);
  });

  test('renders with style "gap" with gap prop with multiple value', () => {
    const gap = [5, 10];
    render(<Space gap={gap}>Test</Space>);

    expect(screen.getByText('Test')).toHaveStyle(`gap: ${gap[0]}px ${gap[1]}px`);
  });
});
