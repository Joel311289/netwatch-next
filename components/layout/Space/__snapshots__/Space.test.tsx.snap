// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Space component renders component correctly 1`] = `
<div>
  <div
    class="spaceWrapper wrapper"
  >
    Test
    <style>
      
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          
          
          
        }
      
    </style>
  </div>
</div>
`;
